gloo/transport/uv/pair.h

```diff
@@ -0,0 +1,248 @@
+
+  virtual ~Pair();
+
+  Pair(const Pair& that) = delete;
+
+  Pair& operator=(const Pair& that) = delete;
+
+  virtual const Address& address() const override;
+
+  virtual void connect(const std::vector<char>& bytes) override;
+
+  virtual void setSync(bool sync, bool busyPoll) override {
+    abort();
+  }
+
+  virtual std::unique_ptr<::gloo::transport::Buffer> createSendBuffer(
+      int slot,
+      void* ptr,
+      size_t size) override {
+    abort();
+  }
+
+  virtual std::unique_ptr<::gloo::transport::Buffer> createRecvBuffer(
+      int slot,
+      void* ptr,
+      size_t size) override {
+    abort();
+  }
+
+  // Send from the specified buffer to remote side of pair.
+  void send(
+      transport::UnboundBuffer* tbuf,
+      uint64_t tag,
+      size_t offset,
+      size_t nbytes) override;
+
+  // Receive into the specified buffer from the remote side of pair.
+  void recv(
+      transport::UnboundBuffer* tbuf,
+      uint64_t tag,
+      size_t offset,
+      size_t nbytes) override;
+
+  // Attempt to receive into the specified buffer from the remote side
+  // of pair. Returns true if there was a remote pending send and the
+  // recv is in progress, false otherwise.
+  bool tryRecv(
+      transport::UnboundBuffer* tbuf,
+      uint64_t tag,
+      size_t offset,
+      size_t nbytes);
+
+  void close() override;
+
+ private:
+  std::mutex mutex_;
+  std::condition_variable cv_;
+
+  // Details for pending send/recv operations.
+  struct UnboundBufferOp {
+    UnboundBufferOp(
+        WeakNonOwningPtr<UnboundBuffer> buf,
+        size_t offset,
+        size_t length)
+        : buf(std::move(buf)), offset(offset), length(length) {}
+
+    WeakNonOwningPtr<UnboundBuffer> buf;
+    size_t offset = 0;
+    size_t length = 0;
+  };
+
+  // Keep track of pending send/recv operations by tag.
+  std::unordered_map<uint64_t, std::deque<UnboundBufferOp>> localPendingSend_;
+  std::unordered_map<uint64_t, std::deque<UnboundBufferOp>> localPendingRecv_;
+
+  // Refer to parent context using raw pointer. This could be a
+  // weak_ptr, seeing as the context class is a shared_ptr, but:
+  // 1) That means calling std::weak_ptr::lock() everytime we need it,
+  // 2) The context holds a unique_ptr to this pair, so the context
+  //    pointer will be valid for the lifetime of this pair.
+  Context* const context_;
+
+  // Refer to device using raw pointer. The context owns a shared_ptr
+  // to the device, and per the lifetime guarantees of the context,
+  // there is no need to duplicate that shared_ptr in this class.
+  Device* const device_;
+
+  // Rank of the process this pair connects to.
+  const int rank_;
+
+  // Timeout for operations executed against this pair.
+  const std::chrono::milliseconds timeout_;
+
+  // Address that the remote side of this pair can connect to.
+  Address addr_;
+
+  // Handle of the connection, if the connection was made successfully.
+  std::shared_ptr<libuv::TCP> handle_;
```

Comment:
There is one pair per rank in a single context. A pair represents the underlying connection between two ranks and all of the associated runtime state (for example all pending or in flight send and recv operations).

It's a `shared_ptr` because of how the libuv wrapper works. I modeled it after `uvw` where every libuv resource is encapsulated in a shared_ptr. This is done because it doesn't force you to maintain ownership of a resource. Instead, you can create one, register a few lambdas for callbacks, issue a request, and then discard it again. Ownership is then passed to libuv by leaking the shared pointer upon making the call and "unleaking" it when the callback is made. If you do want to keep it around yourself as well, you'll need shared ownership.